### ISSUE TITLE
trying a more fair training/testing split like 80/20

### ISSUE THREAD

**Issue by realmarcin** [monarch-initiative/N2V#90]
@vidarmehr would it be possible for you to try a 80/20 training/testing split for your link prediction runs? This will help avoid overfitting and please reviewers, and help us compare to the BioNEV paper.

We missed you on the call today -- some of the discussion is pasted in gitter so you can take a look.

**Comment by vidarmehr** [comment]
@realmarcin Hi Marcin,
Sure. I was going to run new simulations. So, I will chose 80/20 training/testing split this time.
By the way, which BioNEV paper are you referring to? 

Sorry about that. I had a medial appointment. Thank you for update. I will follow the discussion on gitter.

**Comment by justaddcoffee** [comment]
Marcin is referring to this paper:

[Graph Embedding on Biomedical Networks: Methods, Applications, and Evaluations.
Yue et al, Bioinformatics 2019.](https://www.ncbi.nlm.nih.gov/pubmed/31584634)

They have a gitrepo for this paper [here](https://github.com/xiangyue9607/BioNEV)

We reviewed this paper in a journal club yesterday, and a lot of possibly useful suggestions came out of this. 

Two other possible suggestions that came out of this discussion:
- Table 5 of this paper suggests that the optimal walk length and num of walks should be really high (64 - 256), and also that optimal `p` and `q` might be  < 0.25 depending on graph
- could consider rerunning the experiment using only physical protein-protein interactions (if I understand correctly, the experiment currently considers all protein-protein associations, include genetic interactions, etc).

**Comment by vidarmehr** [comment]
@justaddcoffee Thank you very much, Justin! I will read the paper and rerun link prediction with their suggested parameters.

**Comment by vidarmehr** [comment]
I have used 80/20 training/testing split for the link prediction and running simulations on both the entire graph and max. connected component.

**Comment by vidarmehr** [comment]
@pnrobinson 
Hi everyone,
Thank you all @callahantiff @LucaCappelletti94 @justaddcoffee @realmarcin for your helpful suggestions and discussions yesterday at our zoom meeting.  Regarding implementing cross validation on graphs, I have two problems to implement it:
1. How to use sklearn cross validation methods so that the training set consists of all nodes of the graph? Remember that we need all nodes of the graph in the training because the nodes of the training set (which consists of 80% of edges and 80% of non-edges) is going to be embedded first. So, all nodes of the graph must be in the training set so that we have embeddings of all nodes of the graph. 
2. Either I consider the largest component of the graph or the entire graph, I should make sure that an edge which is removed (to be considered in a test set) does not create a new component or does not create nodes with degree 0.

For example:
`from sklearn.datasets import load_iris`
`from sklearn.linear_model import LogisticRegressionCV`
`X, y = load_iris(return_X_y=True)`
`clf = LogisticRegressionCV(cv=5, random_state=0).fit(X, y)`

implements logistic regression with 5 fold cv, but does not check the 2 cases that I mentioned above. 

Do you have any ideas? Should I just use sklearn methods for CV and check if the training set consists of all nodes and no new component is made by edge removal?

What do you think if instead of CV, I just create training/validation/test edges (70%,20%,10%) ?
Thank you.

**Comment by callahantiff** [comment]
> @pnrobinson
> Hi everyone,
> Thank you all @callahantiff @LucaCappelletti94 @justaddcoffee @realmarcin for your helpful suggestions and discussions yesterday at our zoom meeting. Regarding implementing cross validation on graphs, I have two problems to implement it:
> 
> 1. How to use sklearn cross validation methods so that the training set consists of all nodes of the graph? Remember that we need all nodes of the graph in the training because the nodes of the training set (which consists of 80% of edges and 80% of non-edges) is going to be embedded first. So, all nodes of the graph must be in the training set so that we have embeddings of all nodes of the graph.
> 2. Either I consider the largest component of the graph or the entire graph, I should make sure that an edge which is removed (to be considered in a test set) does not create a new component or does not create nodes with degree 0.
> 
> For example:
> `from sklearn.datasets import load_iris`
> `from sklearn.linear_model import LogisticRegressionCV`
> `X, y = load_iris(return_X_y=True)`
> `clf = LogisticRegressionCV(cv=5, random_state=0).fit(X, y)`
> 
> implements logistic regression with 5 fold cv, but does not check the 2 cases that I mentioned above.
> 
> Do you have any ideas? Should I just use sklearn methods for CV and check if the training set consists of all nodes and no new component is made by edge removal?
> 
> What do you think if instead of CV, I just create training/validation/test edges (70%,20%,10%) in the same way that I have already created training/test sets?
> Thank you.

Hi Vida! I see now why you can't do CFV. Is there anyway that you can first embed the whole graph and then perform CFV on on the embedded nodes? If not, I think training/validation/test edges (70%,20%,10%), assuming you run several iterations where you are drawing a new random training set (not touching the validation set once it has initially been drawn) is reasonable.

I'll keep looking through the literature to see if I can help find anything else too.

**Comment by pnrobinson** [comment]
I think we had a function that removes a certain number of edges from a connected component (CC) and then checks that the CC was not disconnected. We should probably add the check that the number of nodes is not changed. We used networks for this.
We could do this separately for each of the CC in the graph (but we might want to remove CCs with less than a threshold number of nodes?).
Probably it is fairest to take positive training/test edges in proportion to the relative sizes of the CCs.
I have not does something like this before and the articles I have looked at say surprisingly little about this topic.

**Comment by vidarmehr** [comment]
> > @pnrobinson
> > Hi everyone,
> > Thank you all @callahantiff @LucaCappelletti94 @justaddcoffee @realmarcin for your helpful suggestions and discussions yesterday at our zoom meeting. Regarding implementing cross validation on graphs, I have two problems to implement it:
> > 
> > 1. How to use sklearn cross validation methods so that the training set consists of all nodes of the graph? Remember that we need all nodes of the graph in the training because the nodes of the training set (which consists of 80% of edges and 80% of non-edges) is going to be embedded first. So, all nodes of the graph must be in the training set so that we have embeddings of all nodes of the graph.
> > 2. Either I consider the largest component of the graph or the entire graph, I should make sure that an edge which is removed (to be considered in a test set) does not create a new component or does not create nodes with degree 0.
> > 
> > For example:
> > `from sklearn.datasets import load_iris`
> > `from sklearn.linear_model import LogisticRegressionCV`
> > `X, y = load_iris(return_X_y=True)`
> > `clf = LogisticRegressionCV(cv=5, random_state=0).fit(X, y)`
> > implements logistic regression with 5 fold cv, but does not check the 2 cases that I mentioned above.
> > Do you have any ideas? Should I just use sklearn methods for CV and check if the training set consists of all nodes and no new component is made by edge removal?
> > What do you think if instead of CV, I just create training/validation/test edges (70%,20%,10%) in the same way that I have already created training/test sets?
> > Thank you.
> 
> Hi Vida! I see now why you can't do CFV. Is there anyway that you can first embed the whole graph and then perform CFV on on the embedded nodes? If not, I think training/validation/test edges (70%,20%,10%), assuming you run several iterations where you are drawing a new random training set (not touching the validation set once it has initially been drawn) is reasonable.
> 
> I'll keep looking through the literature to see if I can help find anything else too.

Thank you, Tiffany!
 No! The entire graph cannot be embedded first, because if the entire graph is embedded first before removing edges, we are including information about the test set in the embedding of the graph (data leakage), while we would like to predict test edges using training edges and embedding of training sets only.

**Comment by LucaCappelletti94** [comment]
So the selection of edges was not truly random but limited to the edges of nodes that have at least another edge, right?

**Comment by vidarmehr** [comment]
> So the selection of edges was not truly random but limited to the edges of nodes that have at least another edge, right?

When I considered just the largest component of the graph and performed edge removal on this subgraph, every time that I removed an edge, I made sure that the graph is still connected.

I am currently working on creating training/test sets from the entire graph. I first removed edges randomly, but I got an error when I performed link prediction, which I assume the error is related to existing some nodes that have degree 0 in the graph. So, I am going to put a condition to make sure that all nodes have at least degree 1. We may even need a stronger condition which is check if no new component is created.

**Comment by LucaCappelletti94** [comment]
How big is the graph? Just to understand how costly is to run the check.

**Comment by vidarmehr** [comment]
> I think we had a function that removes a certain number of edges from a connected component (CC) and then checks that the CC was not disconnected. We should probably add the check that the number of nodes is not changed. We used networks for this.
> We could do this separately for each of the CC in the graph (but we might want to remove CCs with less than a threshold number of nodes?).
> Probably it is fairest to take positive training/test edges in proportion to the relative sizes of the CCs.
> I have not does something like this before and the articles I have looked at say surprisingly little about this topic.

Thank you, Peter. Yes! We checked if removing an edge from the largest connected component, does not make the graph disconnected. So, I think we should do it for every connected component of the graph, when we work with the entire graph.

**Comment by pnrobinson** [comment]
We could probably be more efficient at edge removal by just checking that the affected nodes have a degree of at least 2 before removing an edge?

**Comment by vidarmehr** [comment]
> How big is the graph? Just to understand how costly is to run the check.

The entire graph has almost 10,000 nodes and almost 290,000 edge.

**Comment by vidarmehr** [comment]
> We could probably be more efficient at edge removal by just checking that the affected nodes have a degree of at least 2 before removing an edge?

Yes. I think it is more efficient to check the degree of nodes before edge removal.

**Comment by LucaCappelletti94** [comment]
> > How big is the graph? Just to understand how costly is to run the check.
> 
> The entire graph has almost 10,000 nodes and almost 290,000 edge.

I will ask to a professor of mine that handles truly big graphs, maybe he has faced this issue before.

**Comment by LucaCappelletti94** [comment]
Paolo Boldi, the professor I mentioned in the previous message, suggested we could use [EvalINE](https://arxiv.org/abs/1901.09691). What do you think?

**Comment by pnrobinson** [comment]
This looks quite useful!

**Comment by vidarmehr** [comment]
I will try EvallNE. Currently, I am rerunning simulations for train/test (80/20) and also train/validation/test (70/15/15).

**Comment by justaddcoffee** [comment]
I think we can close this ticket @realmarcin?

**Comment by vidarmehr** [comment]
@justaddcoffee  I think we can close this ticket. train/test (80/20) and also train/validation/test (70/15/15) have been implemented in node2vec-eval (https://github.com/monarch-initiative/node2vec-eval)

**Comment by justaddcoffee** [comment]
+1 @vidarmehr
